File: 查找/153.java
// 使用二分查找寻找换位点，最小元素一定在换位点处
class Solution {
    public int findMin(int[] nums) {
        if (nums.length == 1) {
            return nums[0];
        }
        int left = 0, right = nums.length - 1;
        int mid = 0;
        if (nums[right] > nums[left]) {
            return nums[0];
        }
        while (left <= right) {
            mid = (left + right) / 2;
            // 换位点
            if (nums[mid] > nums[mid + 1]) {
                return nums[mid + 1];
            }
            if (nums[mid] < nums[mid - 1]) {
                return nums[mid];
            }
            // 右边有序，在左侧找换位点
            if (nums[right] > nums[mid]) {
                right = mid - 1;
            }
            // 左边有序，在右侧找换位点
            if (nums[mid] > nums[left]) {
                left = mid + 1;
            }
        }
        return nums[left];
    }
}

File: 查找/153.go
// 交换点（交换位置右边的点）为最小元素，因此本题要求寻找交换点
// 情况1：在交换点一侧的子数组，子数组最左元素 < 子数组最右元素
// 情况2：跨越交换点的子数组，子数组最左元素 > 子数组最右元素
func findMin(nums []int) int {
    left := 0
    right := len(nums) - 1
    for left < right {
        mid := left + (right - left) / 2
        // 情况1：交换点在 mid 左侧（也可能就是mid），之后到左侧寻找（包含mid）
        if nums[mid] < nums[right] {
            right = mid
        } else {
            // 情况2：交换点在mid右侧，之后在右侧寻找（不包含mid）
            left = mid + 1
        }
    }
    return nums[left]
}
